### PR TITLE
feat(queue): implement PoC version of session ↔️ shared group leader interaction

### PR DESCRIPTION
**Pull request by savonarola**
Fixes [EMQX-12306](https://emqx.atlassian.net/browse/EMQX-12306)

Release version: e5.8.0

[EMQX-12306]: https://emqx.atlassian.net/browse/EMQX-12306?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/emqx/src/emqx_persistent_session_ds/shared_subs_agent.hrl [hunk 1]
```diff
@@ -10,10 +10,15 @@
...
-% -define(shared_subs_agent, emqx_ds_shared_sub_agent).
+
+-ifdef(TEST).
+-define(shared_subs_agent, emqx_ds_shared_sub_agent).
+-else.
 %% Till full implementation we need to dispach to the null agent.
 %% It will report "not implemented" error for attempts to use shared subscriptions.
 -define(shared_subs_agent, emqx_persistent_session_ds_shared_subs_null_agent).
+%% -ifdef(TEST).
+-endif.
```
[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx/src/emqx_persistent_session_ds/shared_subs_agent.hrl:21]: It was like you suggested, but that's the formatter's opinion :(

File: apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_proto.hrl [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+-define(agent_connect_leader_msg, agent_connect_leader).
+-define(agent_update_stream_states_msg, agent_update_stream_states).
+-define(agent_connect_leader_timeout_msg, agent_connect_leader_timeout).
+-define(agent_renew_stream_lease_timeout_msg, agent_renew_stream_lease_timeout).
+
+%% Agent messages sent to the leader.
+%% Leader talks to many agents, `agent` field is used to identify the sender.
+
+-define(agent_connect_leader(Agent, TopicFilter), #{
```
[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_proto.hrl:27]: I agree, that's why we have `apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_proto.erl`
We will have BPAPI protocol module around it. 

But inside the remote call we still need some data structures to pass events/messages down the session to the agents.

The intended flow is

Agent -> call to `emqx_ds_shared_sub_proto_v1:connect_leader(...)` -> rpc -> `emqx_ds_shared_sub_proto:connect_leader(...)` -> wrap into `?agent_connect_leader(...)` -> send to Leader _locally_.

[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_proto.hrl:27]: My bad, the `hrl's` lines about sending through network are already irrelevant.

File: apps/emqx/src/emqx_persistent_session_ds/shared_subs_agent.hrl (hunk 1 above)
[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx/src/emqx_persistent_session_ds/shared_subs_agent.hrl]: ✅

File: apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_proto.hrl (hunk 2 above)
[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_proto.hrl:26]: ✅

File: apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_proto.erl [hunk 3]
```diff
@@ -0,0 +1,72 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2024 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+%% TODO
```
[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_proto.erl]: ✅

File: apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub.hrl [hunk 4]
```diff
@@ -0,0 +1,10 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2024 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+-ifndef(EMQX_DS_SHARED_SUB_HRL).
```
[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub.hrl]: ✅

File: apps/emqx/src/emqx_persistent_session_ds/shared_subs_agent.hrl (hunk 1 above)
[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx/src/emqx_persistent_session_ds/shared_subs_agent.hrl]: ✅

File: apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_shared_subs.erl [hunk 5]
```diff
@@ -107,17 +98,26 @@ on_unsubscribe(SessionId, TopicFilter, S0, #{agent := Agent0} = SharedSubS0) ->
...
+    StreamLeaseEvents =/= [] andalso
         ?SLOG(
-            info, #{msg => shared_subs_new_stream_leases, stream_leases => NewLeasedStreams}
+            info, #{
+                msg => shared_subs_new_stream_lease_events, stream_lease_events => StreamLeaseEvents
+            }
         ),
-    S1 = lists:foldl(fun accept_stream/2, S0, NewLeasedStreams),
-    S2 = lists:foldl(fun revoke_stream/2, S1, RevokedStreams),
+    % StreamLeaseEvents =/= [] andalso ct:print("StreamLeaseEvents: ~p~n", [StreamLeaseEvents]),
```
Comment (ieQu1) [inline apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_shared_subs.erl:110]: Remove or replace with a debug log or a tracepoint?

**Comment by ieQu1** [comment]
Maybe I overlooked it, but since agent and leader talk with async messages, shouldn't they monitor each other?

**Comment by savonarola** [comment]
@ieQu1 The protocol specifies timeouts on which agent should reinitialize:

e.g
* `find_leader_timeout ` in the [`connecting` state](https://github.com/emqx/eip/blob/main/active/0028-durable-shared-subscriptions.md#connecting-state)
* `renew_stream_lease_timeout` in the [`replaying` state](https://github.com/emqx/eip/blob/main/active/0028-durable-shared-subscriptions.md#replaying-state)

**Review by ieQu1 — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-3398)

**Pull request merged**